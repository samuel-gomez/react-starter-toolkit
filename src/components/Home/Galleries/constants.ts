import galleries from './galleries.json';

const GALLERIES = Object.values(galleries);

export default GALLERIES;
