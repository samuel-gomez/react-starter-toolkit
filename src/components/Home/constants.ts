export const TITLE_BAR = 'Accueil';
export const TITLE = 'Bienvenue sur la d√©mo du starter Slash Design System';
export const ROUTE_URL_HOME = '/';
