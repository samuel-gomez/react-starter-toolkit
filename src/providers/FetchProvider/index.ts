export { default, FetchContext } from './FetchProvider';
export type { TResponse } from './FetchProvider';
export { default as setResponseError } from './setResponseError';
