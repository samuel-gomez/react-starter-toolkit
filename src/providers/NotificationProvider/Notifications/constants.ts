export const DEFAULT_CLASS_NOTIFICATION = 'af-notifications';
export const MODIFIER_OPEN = 'open';
export const MODIFIER_HIDE = 'hide';
export const DELAY_HIDE = 500;

export const ALERT_ICON = {
  success: 'ok',
  error: 'minus-sign',
  danger: 'alert',
};
