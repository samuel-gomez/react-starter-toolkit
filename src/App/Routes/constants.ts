import { ROUTE_URL_HOME as HOME } from 'pages/Home/constants';
import { ROUTE_URL_SLASH as SLASH } from 'pages/Demos/SlashDesignSystem/constants';
import { ROUTE_URL_NOTFOUND as NOTFOUND } from 'pages/NotFound/constants';
import { ROUTE_URL_UNAUTHORIZE as UNAUTHORIZE } from 'pages/Unauthorize/constants';
import { ROUTE_URL_MEMBERS as MEMBERS } from 'pages/Demos/Members/constants';
import { ROUTE_URL_SEARCHMEMBERS as SEARCHMEMBERS } from 'pages/Demos/SearchMembers/constants';
import { ROUTE_URL_MODAL as MODAL } from 'pages/Demos/Modal/constants';
import { ROUTE_URL_BUTTON as BUTTON } from 'pages/Demos/Button/constants';
import { ROUTE_URL_NOTIFICATION as NOTIFICATION } from 'pages/Demos/Notification/constants';
import { ROUTE_URL_LAYOUT as LAYOUT } from 'pages/Demos/Layout/constants';
import { ROUTE_URL_ALERT as ALERT } from 'pages/Demos/Alert/constants';
import { ROUTE_URL_TEXT_INPUT as TEXT_INPUT } from 'pages/Demos/TextInput/constants';
import { ROUTE_URL_TEXTAREA_INPUT as TEXTAREA_INPUT } from 'pages/Demos/TextareaInput/constants';
import { ROUTE_URL_TABS as TABS } from 'pages/Demos/Tabs/constants';
import { ROUTE_URL_RADIO_INPUT as RADIO_INPUT } from 'pages/Demos/RadioInput/constants';
import { ROUTE_URL_CHECKBOX_INPUT as CHECKBOX_INPUT } from 'pages/Demos/CheckboxInput/constants';
import { ROUTE_URL_SELECT_INPUT as SELECT_INPUT } from 'pages/Demos/SelectInput/constants';
import { ROUTE_URL_STEPPER as STEPPER } from 'pages/Demos/Stepper/constants';
import { ROUTE_URL_RESTITUTION as RESTITUTION } from 'pages/Demos/Restitution/constants';
import { ROUTE_URL_BADGE as BADGE } from 'pages/Demos/Badge/constants';
import { ROUTE_URL_TITLE as TITLE } from 'pages/Demos/Title/constants';
import { ROUTE_URL_HELP as HELP } from 'pages/Demos/Help/constants';
import { ROUTE_URL_HEADER as HEADER } from 'pages/Demos/Header/constants';
import { ROUTE_URL_TABLE as TABLE } from 'pages/Demos/Table/constants';
import { ROUTE_URL_TITLEBAR as TITLEBAR } from 'pages/Demos/TitleBar/constants';
import { ROUTE_URL_ACTION as ACTION } from 'pages/Demos/Action/constants';
import { ROUTE_URL_INFOS as INFOS } from 'pages/Demos/Infos/constants';
import { ROUTE_URL_NAVBAR as NAVBAR } from 'pages/Demos/NavBar/constants';
import { ROUTE_URL_POPOVER as POPOVER } from 'pages/Demos/Popover/constants';
import { ROUTE_URL_FOOTER_CLIENT as FOOTER_CLIENT } from 'pages/Demos/FooterClient/constants';
import { ROUTE_URL_FOOTER as FOOTER } from 'pages/Demos/Footer/constants';
import { ROUTE_URL_ACCORDION as ACCORDION } from 'pages/Demos/Accordion/constants';
import { ROUTE_URL_LOADER as LOADER } from 'pages/Demos/Loader/constants';
import { ROUTE_URL_SLIDER as SLIDER } from 'pages/Demos/Slider/constants';
import { ROUTE_URL_SWITCH as SWITCH } from 'pages/Demos/Switch/constants';

const DEMOS = '/demos';
const LAYOUTS = '/layouts';

const ROUTE_URL = {
  HOME,
  SLASH,
  NOTFOUND,
  UNAUTHORIZE,
  DEMOS,
  MEMBERS,
  SEARCHMEMBERS,
  MODAL,
  BUTTON,
  NOTIFICATION,
  LAYOUTS,
  LAYOUT,
  ALERT,
  TEXT_INPUT,
  TEXTAREA_INPUT,
  TABS,
  RADIO_INPUT,
  CHECKBOX_INPUT,
  SELECT_INPUT,
  RESTITUTION,
  BADGE,
  TITLE,
  HELP,
  HEADER,
  TABLE,
  TITLEBAR,
  ACTION,
  INFOS,
  NAVBAR,
  POPOVER,
  FOOTER_CLIENT,
  FOOTER,
  ACCORDION,
  LOADER,
  SLIDER,
  SWITCH,
  STEPPER,
};

export default ROUTE_URL;
