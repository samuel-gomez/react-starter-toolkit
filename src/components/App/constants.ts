import { ADMIN } from 'shared/constants';

// these mock functions are used when you disable the OIDC authentication
export const useOidcUserFnMock = () => ({ oidcUser: { name: 'John Doe', member_of: [ADMIN] } });
export const useOidcAccessTokenFnMock = () => ({ accessToken: 'token' });
