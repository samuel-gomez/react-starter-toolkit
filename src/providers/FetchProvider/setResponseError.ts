import { STATUS_API, STATUS_HTTP_MESSAGES } from 'shared/constants';

type TsetResponseError = {
  response: {
    anomaly?: {
      label?: string;
      detail?: string;
    };
    status: number;
  };
};

const setResponseError = ({ response }: TsetResponseError) => {
  const { anomaly, status } = response;
  switch (true) {
    case `${status}`.startsWith(`${STATUS_API.WARNING}`):
      return {
        ...anomaly,
        label: anomaly?.label ?? STATUS_HTTP_MESSAGES[status],
        detail: anomaly?.detail ?? '',
        type: 'danger',
        iconName: 'alert',
      };
    case `${status}`.startsWith(`${STATUS_API.ERROR}`):
      return {
        ...anomaly,
        label: anomaly?.label ?? STATUS_HTTP_MESSAGES[status],
        detail: anomaly?.detail ?? '',
      };
    default:
      return {};
  }
};

export default setResponseError;
