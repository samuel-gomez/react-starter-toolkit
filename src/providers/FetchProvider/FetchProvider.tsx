import { createContext, ReactNode, useMemo } from 'react';
import { QueryClient, QueryClientProvider, QueryKey } from '@tanstack/react-query';
import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
import { mergeObj, manageConfig } from 'shared/helpers';
import { API_URL, STATUS_API, STATUS_HTTP_MESSAGES } from 'shared/constants';
import { useOidcAccessToken } from '@axa-fr/react-oidc/dist/ReactOidc';
import fetch from 'cross-fetch';
import setResponseError from './setResponseError';

export type FetchContextType = {
  fetchCustom: (path: string, customConfig: object) => Promise<Response> | null;
  queryClient: QueryClient;
};

export const FetchContext = createContext<FetchContextType | object>({});
FetchContext.displayName = 'FetchContext';

export type TFetchCustom = {
  apiUrl: Record<string, string>;
  fetchAuthConfig: object;
  fetchFn?: typeof fetch;
  mergeObjFn?: typeof mergeObj;
  manageConfigFn?: typeof manageConfig;
};

export type TResponse = {
  status: number;
  blob: () => Promise<unknown>;
  text: () => Promise<unknown>;
  json: () => Promise<object>;
};

type TConfig = {
  blob: boolean;
  text: boolean;
};

export const computeDataError = async (response: TResponse, setResponseErrorFn = setResponseError) => {
  try {
    const data = await response.json();
    return setResponseErrorFn({ response: { ...data, status: response.status } });
  } catch (error) {
    return setResponseErrorFn({ response: { anomaly: { label: STATUS_HTTP_MESSAGES[response.status] }, status: response.status } });
  }
};

export const buildResponse = async (response: TResponse, config: TConfig, computeDataErrorFn = computeDataError) => {
  const { status } = response;
  switch (true) {
    case `${status}`.startsWith(`${STATUS_API.ERROR}`):
    case `${status}`.startsWith(`${STATUS_API.WARNING}`): {
      throw await computeDataErrorFn(response);
    }
    case `${status}` === `${STATUS_API.SUCCESS}`:
      if (config.blob) {
        return response.blob();
      }
      if (config.text) {
        return response.text();
      }
      return {
        ...(await response.json()),
        statusHttp: status,
      };
    default:
      return {
        statusHttp: status,
      };
  }
};

export const setFetchCustom =
  ({ apiUrl, fetchAuthConfig, fetchFn = fetch, mergeObjFn = mergeObj, manageConfigFn = manageConfig }: TFetchCustom) =>
  async (queryKey: QueryKey) => {
    const [path, customConfig, apiName = API_URL.BASE]: QueryKey = queryKey;
    const url = `${apiUrl[apiName as string]}${path}`;
    const fetchAuthConfigCustom = manageConfigFn(apiName as string, fetchAuthConfig);
    const config = mergeObjFn(fetchAuthConfigCustom, customConfig);
    const response = await fetchFn(url, config);
    return buildResponse(response, config as TConfig);
  };

export type TFetchProvider = Omit<TFetchCustom, 'fetchFn' | 'fetchAuthConfig'> & {
  fetchConfig: object;
  children: ReactNode;
  useOidcAccessTokenFn: typeof useOidcAccessToken;
  setQueryClientFn?: typeof setQueryClient;
  setFetchCustomFn?: typeof setFetchCustom;
  showReactQueryDevtoolsComponent?: (process: string | undefined) => JSX.Element | boolean;
};

export const defaultQueryWithAuth = async (queryKey: QueryKey, fetchCustom: ReturnType<typeof setFetchCustom>) => fetchCustom(queryKey);

export const showReactQueryDevtools = (process: string | undefined) =>
  process === 'development' && <ReactQueryDevtools initialIsOpen={false} position="bottom-right" />;

type TsetQueryClient = {
  fetchCustom: ReturnType<typeof setFetchCustom>;
  setQueryFn?: typeof setQuery;
};

export const setQuery =
  (fetchCustom: ReturnType<typeof setFetchCustom>) =>
  ({ queryKey, defaultQueryWithAuthFn = defaultQueryWithAuth }: { queryKey: QueryKey; defaultQueryWithAuthFn?: typeof defaultQueryWithAuth }) =>
    defaultQueryWithAuthFn(queryKey, fetchCustom);

export const setQueryClient = ({ fetchCustom, setQueryFn = setQuery }: TsetQueryClient) =>
  new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        queryFn: setQueryFn(fetchCustom),
      },
    },
  });

const FetchProvider = ({
  apiUrl,
  fetchConfig,
  children,
  useOidcAccessTokenFn,
  mergeObjFn = mergeObj,
  setFetchCustomFn = setFetchCustom,
  showReactQueryDevtoolsComponent = showReactQueryDevtools,
  setQueryClientFn = setQueryClient,
}: TFetchProvider) => {
  const { accessToken } = useOidcAccessTokenFn();

  const authConfig = {
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  };
  const fetchAuthConfig = mergeObjFn(fetchConfig, authConfig);
  const fetchCustom = setFetchCustomFn({ apiUrl, fetchAuthConfig });
  const queryClient = setQueryClientFn({ fetchCustom });
  const value = useMemo(() => ({ fetchCustom, queryClient }), [fetchCustom, queryClient]);

  return (
    <FetchContext.Provider value={value}>
      <QueryClientProvider client={queryClient}>
        {children}
        {showReactQueryDevtoolsComponent(process.env.NODE_ENV)}
      </QueryClientProvider>
    </FetchContext.Provider>
  );
};
export default FetchProvider;
