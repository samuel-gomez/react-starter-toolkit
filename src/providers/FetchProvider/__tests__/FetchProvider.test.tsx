import { useContext } from 'react';
import { render } from '@testing-library/react';
import { STATUS_HTTP_MESSAGES } from 'shared/constants';
import { QueryKey } from '@tanstack/react-query';
import FetchProvider, {
  buildResponse,
  FetchContext,
  setFetchCustom,
  computeDataError,
  showReactQueryDevtools,
  setQueryClient,
  setQuery,
  defaultQueryWithAuth,
} from '../FetchProvider';

describe('defaultQueryWithAuth', () => {
  it('Should call fetchCustom with querykey when defaultQueryWithAuthFn have been called with fetchcustom and querykey', async () => {
    const fetchCustom = jest.fn();
    defaultQueryWithAuth('queryKey' as unknown as QueryKey, fetchCustom);
    expect(fetchCustom).toHaveBeenCalledWith('queryKey');
  });
});

const fetchConfigMock = {
  headers: {
    'Content-Type': 'text/plain',
  },
};

const apiMock = {
  base: 'http://localhost:5001/api/',
  github: 'https://raw.githubusercontent.com/',
};

type TBase = {
  fetchCustom?: (path: string, customConfig: object) => Promise<Response> | null;
};

const Base = ({ fetchCustom }: TBase) => <div>{fetchCustom ? 'haveFetchCustom' : 'notHaveFetchCustom'}</div>;

const BaseWithFetch = () => {
  const fetchProps = useContext(FetchContext);
  return <Base {...fetchProps} />;
};

const useOidcAccessTokenMock = jest.fn().mockReturnValue({
  accessToken: 'accessTokenfdsfdsqgvqvsqfs',
});

describe('FetchProvider', () => {
  it('Should Base have fetchCustom props when render FetchProvider with required props', async () => {
    const { asFragment, getByText } = render(
      <FetchProvider apiUrl={apiMock} fetchConfig={fetchConfigMock} useOidcAccessTokenFn={useOidcAccessTokenMock}>
        <BaseWithFetch />
      </FetchProvider>,
    );

    expect(getByText(/haveFetchCustom/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('setQuery', () => {
  it('Should call setQueryFn with fetchcustom and querykey when defaultQueryWithAuthFn have been called with fetchcustom and querykey', async () => {
    const fetchCustom = jest.fn();
    const defaultQueryWithAuthFn = jest.fn();
    setQuery(fetchCustom)({ queryKey: 'queryKey' as unknown as QueryKey, defaultQueryWithAuthFn });

    expect(defaultQueryWithAuthFn).toHaveBeenCalledWith('queryKey', fetchCustom);
  });
  it('Should return when setQueryClient have been called with fetchcustom', async () => {
    const fetchCustom = jest.fn();
    const result = setQuery(fetchCustom)({ queryKey: 'queryKey' as unknown as QueryKey });

    expect(result).toEqual(Promise.resolve({}));
  });
});

describe('setQueryClient', () => {
  it('Should call setQueryFn with fetch custom when setQueryClient have been called with fetchcustom', async () => {
    const fetchCustom = jest.fn();
    const setQueryFn = jest.fn();
    setQueryClient({ fetchCustom, setQueryFn });

    expect(setQueryFn).toHaveBeenCalledWith(fetchCustom);
  });
});

const fetchMock = jest.fn();

describe('setFetchCustom', () => {
  const resolvedValue = { json: () => ({ data: 'data' }), blob: () => ({ blob: 'blob' }), status: 200 };
  it('Should fetch called with merged infos and result return 200 response Fetch when call setFetchCustom', async () => {
    const queryKey = [
      'members',
      {
        headers: {
          body: 'body',
        },
      },
    ];
    fetchMock.mockResolvedValue(resolvedValue);
    const result = await setFetchCustom({ apiUrl: apiMock, fetchAuthConfig: fetchConfigMock, fetchFn: fetchMock })(queryKey);

    expect(fetchMock).toBeCalledWith('http://localhost:5001/api/members', {
      headers: {
        'Content-Type': 'text/plain',
        body: 'body',
      },
    });

    expect(result).toEqual({ data: 'data', statusHttp: 200 });
  });

  it('Should return blob response when config.blob is true', async () => {
    const queryKey = [
      'members',
      {
        blob: true,
      },
    ];
    fetchMock.mockResolvedValue(resolvedValue);
    const result = await setFetchCustom({ apiUrl: apiMock, fetchAuthConfig: fetchConfigMock, fetchFn: fetchMock })(queryKey);
    expect(result).toEqual({ blob: 'blob' });
  });
});

describe('buildResponse', () => {
  const computeDataErrorFn = jest.fn();
  const responseMock = {
    status: 200,
    text: async () => ({ msg: 'testText' }),
    json: async () => ({ msg: 'test' }),
    blob: async () => ({ msg: 'testblob' }),
  };
  it('Should return response 200 and msg = test when status = 200 and blob is falsy', async () => {
    const result = await buildResponse(responseMock, { blob: false, text: false });
    expect(result).toEqual({ statusHttp: responseMock.status, msg: 'test' });
  });

  it('Should return msg = testblob when status = 200 and blob is truthy', async () => {
    const result = await buildResponse(responseMock, { blob: true, text: false });
    expect(result).toEqual({ msg: 'testblob' });
  });

  it('Should return msg = testText when status = 200 and text is truthy', async () => {
    const result = await buildResponse(responseMock, { blob: false, text: true });
    expect(result).toEqual({ msg: 'testText' });
  });

  it('Should called computeDataErrorFn when status = 504', async () => {
    responseMock.status = 504;
    try {
      await buildResponse(responseMock, { blob: false, text: false }, computeDataErrorFn);
    } catch (error) {
      expect(computeDataErrorFn).toBeCalled();
    }
  });

  it('Should called computeDataErrorFn when status = 500', async () => {
    responseMock.status = 500;
    try {
      await buildResponse(responseMock, { blob: false, text: false }, computeDataErrorFn);
    } catch (error) {
      expect(computeDataErrorFn).toBeCalled();
    }
  });

  it('Should called computeDataErrorFn when status = 400', async () => {
    responseMock.status = 400;
    try {
      await buildResponse(responseMock, { blob: false, text: false }, computeDataErrorFn);
    } catch (error) {
      expect(computeDataErrorFn).toBeCalled();
    }
  });

  it('Should return response 204 when status = 204', async () => {
    responseMock.status = 204;
    const result = await buildResponse(responseMock, { blob: false, text: false });
    expect(result).toEqual({ statusHttp: responseMock.status });
  });
});

describe('computeDataError', () => {
  const responseMock = {
    status: 500,
    text: async () => ({ msg: 'test' }),
    json: async () => ({}),
    blob: async () => ({ msg: 'testblob' }),
  };
  const setResponseFn = jest.fn();

  it('Should called setResponseFn without setResponseFn', async () => {
    responseMock.json = async () => ({
      anomaly: { label: 'test' },
      code: 500,
    });
    try {
      await computeDataError(responseMock);
      expect(setResponseFn).toBeCalledWith({ label: 'test' });
    } catch (error) {
      /* empty */
    }
  });

  it('Should called setResponseFn', async () => {
    responseMock.json = async () => ({
      anomaly: { label: 'test' },
      code: 500,
    });
    try {
      await computeDataError(responseMock, setResponseFn);
      expect(setResponseFn).toBeCalledWith({ label: 'test' });
    } catch (error) {
      /* empty */
    }
  });

  it('Should called setResponseFn in throw', async () => {
    responseMock.json = async () => {
      throw new Error('');
    };
    try {
      await computeDataError(responseMock, setResponseFn);
    } catch (error) {
      expect(setResponseFn).toBeCalledWith({
        response: { anomaly: { label: STATUS_HTTP_MESSAGES[responseMock.status] }, status: responseMock.status },
      });
    }
  });
});

describe('showReactQueryDevtools', () => {
  it('Should not return ReactQueryDevTools', () => {
    const process = 'test';
    const result = showReactQueryDevtools(process);
    expect(result).toEqual(false);
  });

  it('Should return ReactQueryDevTools', () => {
    const process = 'development';
    const result = showReactQueryDevtools(process);
    expect(result).not.toEqual(false);
  });
});
