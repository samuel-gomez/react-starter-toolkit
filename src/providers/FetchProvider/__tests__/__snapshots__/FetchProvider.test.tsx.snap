// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchProvider Should Base have fetchCustom props when render FetchProvider with required props 1`] = `
<DocumentFragment>
  <div>
    haveFetchCustom
  </div>
</DocumentFragment>
`;
